mesh/mesh.go

```diff
@@ -181,10 +187,50 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
 			m.With().Error("failed to push transactions", log.Err(err))
 			break
 		}
+		m.logStateRoot(layerId)
+		m.setLayerHash(layerId)
 	}
+	m.persistLayerHash()
 	m.Info("done validating layer %v", lyr.Index())
 }
 
+func (m *Mesh) logStateRoot(layerId types.LayerID) {
+	m.Event().Info("end of layer state root",
+		log.LayerId(layerId.Uint64()),
+		log.String("state_root", util.Bytes2Hex(m.TxProcessor.GetStateRoot().Bytes())),
+	)
+}
+
+func (m *Mesh) setLayerHash(layerId types.LayerID) {
+	validBlockIds, err := m.ContextuallyValidBlock(layerId)
+	if err != nil {
+		m.With().Error("failed to get contextually valid blocks for layer hash - all future hashes will be inconsistent",
+			log.Err(err), log.LayerId(layerId.Uint64()))
+		return
+	}
+	layerHash := sha256.New()
+	layerHash.Write(m.layerHash)
+	sortedBlockIds := make([]types.BlockID, 0, len(validBlockIds))
+	for id := range validBlockIds {
+		sortedBlockIds = append(sortedBlockIds, id)
+	}
+	types.SortBlockIds(sortedBlockIds)
+	for _, id := range sortedBlockIds {
+		layerHash.Write(id.ToBytes())
+	}
+	m.layerHash = layerHash.Sum(nil)
+	m.Event().Info("new layer hash",
+		log.LayerId(layerId.Uint64()),
+		log.String("layer_hash", util.Bytes2Hex(m.layerHash)))
+}
+
+func (m *Mesh) persistLayerHash() {
```

Comment:
does this persist only the last layer hash? if so make sure it is understandable from the func name, however I think we should keep a key value between all layers and respeceted hashes